DSharpPlus/Entities/DiscordPermissions.Enumeration.cs.cs

```diff
@@ -0,0 +1,96 @@
+#pragma warning disable IDE0040
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Numerics;
+
+using CommunityToolkit.HighPerformance.Helpers;
+
+namespace DSharpPlus.Entities;
+
+// contains the implementation details for EnumeratePermissions
+partial struct DiscordPermissions
+{
+    /// <summary>
+    /// Presents a slim enumerable wrapper around a set of permissions.
+    /// </summary>
+    public readonly struct DiscordPermissionEnumerable : IEnumerable<DiscordPermission>
+    {
+        private readonly DiscordPermissionContainer data;
+
+        internal DiscordPermissionEnumerable(DiscordPermissionContainer data)
+            => this.data = data;
+
+        /// <summary>
+        /// Gets an enumerator for the present permission set.
+        /// </summary>
+        public readonly DiscordPermissionEnumerator GetEnumerator() => new(this.data);
+
+        // implementations for IEnumerable<T>, we'd like to not box by default
+        IEnumerator<DiscordPermission> IEnumerable<DiscordPermission>.GetEnumerator() => GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+    }
+
+    /// <summary>
+    /// Represents an enumerator for permission fields within a permission set.
+    /// </summary>
+    public struct DiscordPermissionEnumerator : IEnumerator<DiscordPermission>
+    {
+        private readonly DiscordPermissionContainer data;
+
+        internal DiscordPermissionEnumerator(DiscordPermissionContainer data)
+            => this.data = data;
+
+        public readonly DiscordPermission Current => (DiscordPermission)(this.block << 5) + this.bit;
```

Comment:
Dare I ask the thought process behind this math

Reply:
each block is 32 bits, so `this.block * 32 + this.bit`